snapcraft/parts/plugins/matter_plugin.py

```diff
@@ -0,0 +1,138 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2023 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The matter plugin."""
+import os
+
+from typing import Any, Dict, List, Set
+
+from craft_parts import infos, plugins
+from overrides import overrides
+
+MATTER_REPO = "https://github.com/project-chip/connectedhomeip.git"
+"""The repository where the matter SDK resides."""
+
+
+class MatterPluginProperties(plugins.PluginProperties, plugins.PluginModel):
+    """The part properties used by the matter plugin."""
+
+    # matter_branch: str
+    # zap_version: str
```

Comment:
The Snapcraft could be built using the new Matter plugin and successfully create a snap with hardcoded arguments (`matter_branch` and `zap_version`). However, currently it's not yet possible to pass these two arguments dynamically. Below is an example of how this plugin has been used in the snapcraft.yaml file and the build errors it throws:
```
parts:
  chip-tool:
    plugin: matter
    matter-branch: "v1.2.0.1"
    zap-version: "v2023.11.13"
    override-build: |
...
```
```
$ snap info snapcraft
installed:          6.0.post966+git4eed4be8               (x1) 78MB classic
$ snapcraft -v --debug
...
Bad snapcraft.yaml content:
- extra field 'matter-branch' not permitted in 'parts.chip-tool' configuration 
```

Reply:
One argument in the snapcraft.yaml has been accepted by snapcraft after rebased the working branch onto tags 8.0.0 instead of main branch.

Reply:
Also, rebasing on `8.0.0` shouldn't make a difference vs `main` (there are only a few commits between the two that should be unrelated).  We will have to rebase this PR onto `main` to merge anyways.

On `8.0.0` and `main`, I got the same error at the same place:
```
2023-12-13 12:50:53.026 :: pw_cli.plugins.Error: Failed to import module "pw_watch.watch" (ImportError: cannot import name 'OneStyleAndTextTuple' from 'prompt_toolkit.formatted_text' (/root/snapcraft-examples/matter/parts/chip-tool/build/matter/.environment/pigweed-venv/lib/python3.11/site-packages/prompt_toolkit/formatted_text/__init__.py))
2023-12-13 12:50:53.080 :: ./scripts/setup/bootstrap.sh: line 140: pop_var_context: head of shell_variables not a function context
2023-12-13 12:50:53.080 :: ./scripts/setup/bootstrap.sh: line 153: pop_var_context: head of shell_variables not a function context
2023-12-13 12:50:53.258 'override-build' in part 'chip-tool' failed with code 1.
Review the scriptlet and make sure it's correct.
```

Reply:
Thank you for pointing out the keyword prefix. The updates have been made on both snapcraft and [example snap](https://github.com/MonicaisHer/chip-tool-snap/blob/test-matter-plugin/snap/snapcraft.yaml#L23-L25) sides. However, although snapcraft appears to accept these keywords, it still fails to map `x-y` to `x_y`, resulting in a following error when running snapcraft:
```
# clean install the snapcraft built from this PR
$ sudo snap remove --purge snapcraft
$ sudo snap install snapcraft/snapcraft_8* --dangerous --classic

# build an example snap: 
$ cd chip-tool-snap
$ snapcraft clean
$ snapcraft -v --debug
...
Building chip-tool                                                                                                                                                             
:: + craftctl default                                                                                                                                                          
'MatterPlugin' object has no attribute 'matter_tag'
```

And you are right; the rebase shouldn't matter in this context. It could be caused by cache. Regarding the error you're encountering in `./scripts/setup/bootstrap.sh`, I haven't had a chance to encounter the same error due to the previous error: `MatterPlugin' object has no attribute 'matter_tag'`.

Reply:
Thank you for pointing out the keyword prefix. It is working now! And you are right; the rebase shouldn't matter in this context. It could be caused by cache.

Reply:
The `OneStyleAndTextTuple` error you encountered is a [bug](https://github.com/project-chip/connectedhomeip/issues/30981) from prompt-toolkit. This issue has been [fixed](https://github.com/project-chip/connectedhomeip/pull/30987) in the latest Matter SDK. By using the fix commit and the Matter plugin, example snap can be built successfully now.